carma_cooperative_perception/src/multiple_object_tracker_component.cpp

```diff
@@ -0,0 +1,446 @@
+
+            RCLCPP_ERROR(
+              get_logger(), "Cannot change parameter 'execution_frequencey_hz': " + result.reason);
+
+            break;
+          } else {
+            this->execution_period_ = 1 / units::frequency::hertz_t{parameter.as_double()};
+          }
+        } else {
+          result.successful = false;
+          result.reason = "Unexpected parameter name '" + parameter.get_name() + '\'';
+        }
+      }
+
+      return result;
+    });
+
+  RCLCPP_INFO(get_logger(), "Lifecycle transition: successfully configured");
+
+  return carma_ros2_utils::CallbackReturn::SUCCESS;
+}
+
+auto MultipleObjectTrackerNode::handle_on_activate(
+  const rclcpp_lifecycle::State & /* previous_state */) -> carma_ros2_utils::CallbackReturn
+{
+  RCLCPP_INFO(get_logger(), "Lifecycle transition: actiavting");
+
+  if (pipeline_execution_timer_ == nullptr) {
+    const std::chrono::duration<double, std::nano> period_ns{mot::remove_units(execution_period_)};
+    pipeline_execution_timer_ = create_wall_timer(period_ns, [this] { execute_pipeline(); });
+  } else {
+    pipeline_execution_timer_->reset();
+  }
+
+  RCLCPP_INFO(get_logger(), "Lifecycle transition: successfully activated");
+
+  return carma_ros2_utils::CallbackReturn::SUCCESS;
+}
+
+auto MultipleObjectTrackerNode::handle_on_deactivate(
+  const rclcpp_lifecycle::State & /* previous_state */) -> carma_ros2_utils::CallbackReturn
+{
+  RCLCPP_INFO(get_logger(), "Lifecycle transition: deactivating");
+
+  pipeline_execution_timer_->cancel();
+
+  // There is currently no way to change a timer's period in ROS 2, so we will
+  // have to create a new one in case a user changes the period.
+  pipeline_execution_timer_.reset();
+
+  RCLCPP_INFO(get_logger(), "Lifecycle transition: successfully deactivated");
+
+  return carma_ros2_utils::CallbackReturn::SUCCESS;
+}
+
+auto MultipleObjectTrackerNode::handle_on_cleanup(
+  const rclcpp_lifecycle::State & /* previous_state */) -> carma_ros2_utils::CallbackReturn
+{
+  RCLCPP_INFO(get_logger(), "Lifecycle transition: cleaning up");
+
+  // CarmaLifecycleNode does not handle subscriber pointer reseting for us
+  detection_list_sub_.reset();
+
+  undeclare_parameter("execution_frequency_hz");
+  remove_on_set_parameters_callback(on_set_parameters_callback_.get());
+
+  RCLCPP_INFO(get_logger(), "Lifecycle transition: successfully cleaned up");
+
+  return carma_ros2_utils::CallbackReturn::SUCCESS;
+}
+
+auto MultipleObjectTrackerNode::handle_on_shutdown(
+  const rclcpp_lifecycle::State & /* previous_state */) -> carma_ros2_utils::CallbackReturn
+{
+  RCLCPP_INFO(get_logger(), "Lifecycle transition: shutting down");
+
+  // CarmaLifecycleNode does not handle subscriber pointer reseting for us
+  detection_list_sub_.reset();
+
+  RCLCPP_INFO(get_logger(), "Lifecycle transition: successfully shut down");
+
+  return carma_ros2_utils::CallbackReturn::SUCCESS;
+}
```

Comment:
I think you're assuming this node will be used only within a lifecycle component wrapper instance, which I don't think is accurate. The lifecycle component manager is a different concept from the lifecycle node.

I agree it's tedious to do the same logging messages and cleanup for all derivative nodes. There are open issues(usdot-fhwa-stol/carma-utils#154 and usdot-fhwa-stol/carma-utils#157) to address some of these concerns. Wrapping the subscriptions in a lifecycle-like interfaces requires the most boilerplate, so I would like to also move that to some other area. Thoughts?

Reply:
- Yeah I am always thinking how it works inside carma-platform not standalone by itself. As part of the whole system in practice, suppose a required node is deactivated, it sends system alert to shut down carma-platform anyways, so resetting subscriptions, timers, or parameters seems unnecessary. If it was not required node, it doesn't affect platform anyways. 
- However, adding logs and moving the timer reset to earlier lifecycle state make sense to me. Subscription reset too. If we move all those to the lifecycle functions, then the developers don't have to write these for each nodes in my opinion. 
- Also these loggers probably need the actual name of the node in the log to be useful. Since many of these nodes can run inside a single container. 
-